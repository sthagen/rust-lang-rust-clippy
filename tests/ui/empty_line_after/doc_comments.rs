#![warn(clippy::empty_line_after_outer_attr, clippy::empty_line_after_doc_comments)]

/// Meant to be an
/// inner doc comment
/// for the crate
//~^ empty_line_after_doc_comments

fn first_in_crate() {}

mod m {

    /// Meant to be an
    /// inner doc comment
    /// for the module
    //~^ empty_line_after_doc_comments

    fn first_in_module() {}
}

mod some_mod {
    //! This doc comment should *NOT* produce a warning

    mod some_inner_mod {
        fn some_noop() {}
    }

    /// # Indented
    //~^ empty_line_after_doc_comments

    /// Blank line
    fn indented() {}
}

/// This should produce a warning
//~^ empty_line_after_doc_comments

fn with_doc_and_newline() {}

// This should *NOT* produce a warning
#[crate_type = "lib"]
/// some comment
fn with_no_newline_and_comment() {}

/// This doc comment should produce a warning
//~^ empty_line_after_doc_comments

/** This is also a doc comment and is part of the warning
 */

#[allow(non_camel_case_types)]
#[allow(missing_docs)]
#[allow(dead_code)]
fn three_attributes() {}

mod misattributed {

    /// docs for `old_code`
    //~^ empty_line_after_doc_comments
    // fn old_code() {}

    fn new_code() {}

    /// Docs
    /// for OldA
    //~^ empty_line_after_doc_comments
    // struct OldA;

    /// Docs
    /// for OldB
    // struct OldB;

    /// Docs
    /// for Multiple
    #[allow(dead_code)]
    struct Multiple;
}

mod block_comments {

    /**
     * Meant to be inner doc comment
     */
    //~^^^ empty_line_after_doc_comments

    fn first_in_module() {}

    /**
     * Docs for `old_code`
     */
    //~^^^ empty_line_after_doc_comments
    /* fn old_code() {} */

    /**
     * Docs for `new_code`
     */
    fn new_code() {}

    /// Docs for `old_code2`
    //~^ empty_line_after_doc_comments
    /* fn old_code2() {} */

    /// Docs for `new_code2`
    fn new_code2() {}
}

// This should *NOT* produce a warning
#[doc = "
Returns the escaped value of the textual representation of

"]
pub fn function() -> bool {
    true
}

// This should *NOT* produce a warning
#[derive(Clone, Copy)]
pub enum FooFighter {
    Bar1,

    Bar2,

    Bar3,

    Bar4,
}

/// Should not lint
// some line comment
/// gaps without an empty line
struct LineComment;

/// This should *NOT* produce a warning because the empty line is inside a block comment
/*

*/
pub struct EmptyInBlockComment;

/// This should *NOT* produce a warning
/* test */
pub struct BlockComment;

/// Ignore the empty line inside a cfg_attr'd out attribute
#[cfg_attr(any(), multiline(
    foo = 1

    bar = 2
))]
fn empty_line_in_cfg_attr() {}

fn main() {}
