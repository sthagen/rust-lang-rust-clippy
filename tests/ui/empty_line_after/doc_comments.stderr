error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:5:1
   |
LL | / /// for the crate
LL | |
LL | |
   | |_^
LL |   fn first_in_crate() {}
   |   ------------------- the comment documents this function
   |
   = note: `-D clippy::empty-line-after-doc-comments` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::empty_line_after_doc_comments)]`
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `first_in_crate` comment it out
   |
LL ~ // /// Meant to be an
LL ~ // /// inner doc comment
LL ~ // /// for the crate
   |

error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:14:5
   |
LL | /     /// for the module
LL | |
LL | |
   | |_^
LL |       fn first_in_module() {}
   |       -------------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `first_in_module` comment it out
   |
LL ~     // /// Meant to be an
LL ~     // /// inner doc comment
LL ~     // /// for the module
   |

error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:27:5
   |
LL | /     /// # Indented
LL | |
LL | |
   | |_^
LL |       /// Blank line
LL |       fn indented() {}
   |       ------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `indented` comment it out
   |
LL |     // /// # Indented
   |     ++

error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:34:1
   |
LL | / /// This should produce a warning
LL | |
LL | |
   | |_^
LL |   fn with_doc_and_newline() {}
   |   ------------------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `with_doc_and_newline` comment it out
   |
LL | // /// This should produce a warning
   | ++

error: empty lines after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:44:1
   |
LL | / /// This doc comment should produce a warning
LL | |
LL | |
LL | | /** This is also a doc comment and is part of the warning
LL | |  */
LL | |
   | |_^
...
LL |   fn three_attributes() {}
   |   --------------------- the comment documents this function
   |
   = help: if the empty lines are unintentional remove them
help: if the doc comment should not document `three_attributes` comment it out
   |
LL ~ // /// This doc comment should produce a warning
LL |
LL |
LL ~ /* This is also a doc comment and is part of the warning
   |

error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:57:5
   |
LL | /     /// docs for `old_code`
...  |
LL | |
   | |_^
LL |       fn new_code() {}
   |       ------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `new_code` comment it out
   |
LL |     // /// docs for `old_code`
   |     ++

error: empty lines after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:64:5
   |
LL | /     /// for OldA
...  |
LL | |
   | |_^
...
LL |       struct Multiple;
   |       --------------- the comment documents this struct
   |
   = help: if the empty lines are unintentional remove them
help: if the doc comment should not document `Multiple` comment it out
   |
LL ~     // /// Docs
LL ~     // /// for OldA
LL |
LL |     // struct OldA;
LL |
LL ~     // /// Docs
LL ~     // /// for OldB
   |

error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:80:5
   |
LL | /     /**
LL | |      * Meant to be inner doc comment
LL | |      */
LL | |
LL | |
   | |_^
LL |       fn first_in_module() {}
   |       -------------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `first_in_module` comment it out
   |
LL -     /**
LL +     /*
   |

error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:87:5
   |
LL | /     /**
LL | |      * Docs for `old_code`
LL | |      */
...  |
LL | |
   | |_^
...
LL |       fn new_code() {}
   |       ------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `new_code` comment it out
   |
LL -     /**
LL +     /*
   |

error: empty line after doc comment
  --> tests/ui/empty_line_after/doc_comments.rs:98:5
   |
LL | /     /// Docs for `old_code2`
LL | |
LL | |     /* fn old_code2() {} */
LL | |
   | |_^
LL |       /// Docs for `new_code2`
LL |       fn new_code2() {}
   |       -------------- the comment documents this function
   |
   = help: if the empty line is unintentional remove it
help: if the doc comment should not document `new_code2` comment it out
   |
LL |     // /// Docs for `old_code2`
   |     ++

error: aborting due to 10 previous errors

